### PR TITLE
[clang] Improved isSimpleTypeSpecifier

### LINKED ISSUE #72203: Sema::isSimpleTypeSpecifier should return true for `_Bool` from C99 onwards
Relates to an existing FIXME in code: https://github.com/llvm/llvm-project/blob/main/clang/lib/Sema/SemaDecl.cpp#L131

--

As discussed in https://discourse.llvm.org/t/should-issimpletypespecifier-return-true-for-bool/74873, this would be a handy little addition for some use cases (for example in Swift, which uses it in the clang importer).

I have prepared a patch already and will raise a PR after I've raised this issue, to aid in tracking, and to avoid someone duplicating the work in future.

### PR DESCRIPTION
**Pull request by carlos4242**
- Sema::isSimpleTypeSpecifier return true for _Bool in c99 (currently returns false for _Bool, regardless of C dialect). (Fixes #72203)
- replace the logic with a check for simple types and a proper check for a valid keyword in the appropriate dialect

### CODE REVIEW

**Comment by carlos4242** [comment]
@owenca ... is this better for you?

**Comment by owenca** [comment]
See https://discourse.llvm.org/t/should-issimpletypespecifier-return-true-for-bool/74873/16.

File: clang/lib/Sema/SemaDecl.cpp
Comment (owenca) [inline clang/lib/Sema/SemaDecl.cpp:164]: I would move it up, to just below `kw_bool`.

File: clang/include/clang/Sema/Sema.h
Comment (owenca) [inline clang/include/clang/Sema/Sema.h:2639]:
```suggestion
  bool isSimpleTypeSpecifier(const Token &Tok) const;
```

File: clang/lib/Sema/SemaDecl.cpp
Comment (owenca) [inline clang/lib/Sema/SemaDecl.cpp:135]:
```suggestion
bool Sema::isSimpleTypeSpecifier(const Token &Tok) const {
  switch (Tok.getKind()) {
```

File: clang/lib/Sema/SemaDecl.cpp
Comment (owenca) [inline clang/lib/Sema/SemaDecl.cpp:172]:
```suggestion
    return Tok.getIdentifierInfo()->isKeyword(getLangOpts());
```

File: clang/include/clang/Lex/Token.h
Comment (owenca) [inline clang/include/clang/Lex/Token.h:205]: IMO we don't need this.

**Comment by carlos4242** [comment]
I think unit testing the bug fix to this function (the copy that is in Sema) is going to be nearly impossible.

To prove the point, I tried hard coding the wrong return value for many types just now...

```
  switch (Kind) {
  case tok::kw_bool:
  case tok::kw_short:
  case tok::kw_long:
  case tok::kw___int64:
  case tok::kw___int128:
  case tok::kw_signed:
  case tok::kw_unsigned:
  case tok::kw_char:
  case tok::kw_int:
  case tok::kw_half:
  case tok::kw_float:
  case tok::kw_double:
  case tok::kw___bf16:
  case tok::kw__Float16:
  case tok::kw___float128:
  case tok::kw___ibm128:
  case tok::kw_wchar_t:
    return false;
```

...Then re-ran the full clang/tests suite (18,000+ tests) and all passed as normal. So there's basically very little unit test coverage on this function as it stands. It's not just a question of improving or adding to existing tests. Whoever first wrote this function seems to have not added really enough unit test coverage at the time, and it's only called in two obscure places in Sema. The duplicate copy function in Format is more widely used and would probably be easier to test. But for this PR I think I'll make the changes you suggest, then leave it as is... either it gets approved without new tests or it'll get culled one day as an abandoned PR. I'll leave that up to the maintainers.

**Comment by owenca** [comment]
LGTM, but please rebase and resolve conflicts.

**Comment by carlos4242** [comment]
> LGTM, but please rebase and resolve conflicts.

done, rebased, tests pass locally

**Comment by carlos4242** [comment]
note: build kite is passing on linux but failing windows tests with an error like "mt.exe : general error c101008d: Failed to write the updated manifest to the resource of file "bin\llvm-symbolizer.exe". Operation did not complete successfully because the file contains a virus or potentially unwanted software" ... I'm a bit unsure how to proceed with that. Does it look like a genuine failure I should address or a spurious build quirk?

**Review by owenca — APPROVED** [review]

**Review by AaronBallman — APPROVED** [review]
LGTM; I can't see a reasonable way for us to test this change and it's *almost* an NFC change, so I think it's fine to land without tests given that we're at the very start of the 19.x cycle.

**Comment by AaronBallman** [comment]
Once precommit CI comes back green, I'll merge on your behalf.

**Pull request merged**
